handle tx finalization in controllers instead of provider-engine

kumavis opened the pull request:
involved copying a ton of finalization code out of provider-engine and into metamask : /
since we now broadcast the tx ourself, we dont have the issue of submitting txs from persistence

Fixes #1001

kumavis committed:
background - handle tx finalization in controllers instead of provider-engine

kumavis committed:
Merge branch 'dev' into bug-submitTx

kumavis commented on app/scripts/metamask-controller.js:
```diff
@@ -257,24 +239,29 @@ module.exports = class MetamaskController extends EventEmitter {
+      function removeHandlers() {
+        this.txManager.removeListener(`${txMeta.id}:submitted`, successHandler)
+        this.txManager.removeListener(`${txMeta.id}:rejected`, failHandler)
+      }
```
i reccomend adding a single method for complete/fail:
```
this.txManager.once(`${txMeta.id}:finished`, failHandler)
```
can pass the txId

kumavis commented on app/scripts/metamask-controller.js:
```diff
@@ -257,24 +239,29 @@ module.exports = class MetamaskController extends EventEmitter {
+      // listen for tx completion (success, fail)
+      this.txManager.once(`${txMeta.id}:submitted`, successHandler)
+      this.txManager.once(`${txMeta.id}:rejected`, failHandler)
+      function successHandler(rawTx) {
```
actually we dont care about the rawTx what we need to return is the txHash

kumavis commented on app/scripts/transaction-manager.js:
```diff
@@ -15,11 +14,14 @@ module.exports = class TransactionManager extends EventEmitter {
     this.txHistoryLimit = opts.txHistoryLimit
     this.getSelectedAccount = opts.getSelectedAccount
     this.provider = opts.provider
+    this.query = opts.query
```
remove

kumavis commented on app/scripts/transaction-manager.js:
```diff
@@ -144,46 +145,76 @@ module.exports = class TransactionManager extends EventEmitter {
+        (rawTx, cb) => self.publishTransaction(txId, rawTx, cb),
+      ], (err) => {
+        self.nonceLock.leave()
+        // TODO: move tx to error state
```
move tx to error state ?

kumavis commented on app/scripts/transaction-manager.js:
```diff
@@ -144,46 +145,76 @@ module.exports = class TransactionManager extends EventEmitter {
+  publishTransaction (txId, rawTx, cb) {
+    this.txProviderUtils.publishTransaction(rawTx, (err) => {
+      if (err) return cb(err)
+      this.setTxStatusSubmitted(txId, rawTx)
```
dont need to pass this `rawTx` bc we wont emit it

kumavis commented on app/scripts/transaction-manager.js:
```diff
@@ -144,46 +145,76 @@ module.exports = class TransactionManager extends EventEmitter {
+    })
+  }
+
+  validateTxParams (txParams, cb) {
```
this could be move into tx-utils or provider-engine

kumavis commented on app/scripts/transaction-manager.js:
```diff
@@ -294,13 +334,16 @@ module.exports = class TransactionManager extends EventEmitter {
   //    - `'confirmed'` the tx has been included in a block.
-  _setTxStatus (txId, status) {
+  // "value" is an optional parameter to emit
+  _setTxStatus (txId, status, value) {
```
dont need value
remove comment too

kumavis committed:
background - metamask controller - fix 'this' ref

kumavis committed:
Merge branch 'bug-submitTx' of github.com:MetaMask/metamask-plugin into bug-submitTx

kumavis commented:
this needs QA
actual tx signing in ui is broken

Zanibas commented:
Gas calculation is broken here, because getMaxTxAndFee isn't ever called, per @frankiebee

frankiebee commented:
working on it

frankiebee committed:
fix the maxcost not being included in txMeta

frankiebee committed:
Satisfy review needs:
removed unnecessary this.query = opts.query from constructor
Created a tx error state for errors in approveTransaction
validateTxParams has been moved to tx-utils
removed "value" arg from _setTxStatus

frankiebee committed:
fix for linting

frankiebee commented:
woops one more thing

frankiebee committed:
Fix the inclusion of the txFee in the meta tx object

frankiebee commented:
Fixed tx rendering of values Should be good too go!

kumavis commented:
we may be returning the tx hash in the wrong format

kumavis commented:
ok yeah we are not returning the txHash - im on it

kumavis committed:
background - return txHash to provider-engine on done

kumavis committed:
background - txManager - filter txs by network

kumavis committed:
eth-store - emit update on new account add

kumavis commented:
these commits should fix:
* a tx submission bug (returned raw instead of hash)
* metamask badge sometimes shows wrong number (was including txs from different network)
* metamask popup doesnt show balances until next block

kumavis committed:
tests - txManager - add getNetwork fn to fixture txManager

kumavis committed:
test - txManager - add network to all txs

kumavis committed:
test - clear localStorage in case CI caches it (?)

kumavis committed:
test - clear localStorage on test start
